### DISCUSSION TITLE
eta for 0.4.0?

### DISCUSSION THREAD

**Discussion by NumseBacon** [trackmania-rl/tmrl#27] [Q&A] [answered]
title :)

**Comment by yannbouteiller** [comment]
It is coming :D 

0.4.0 's most important feature is that it now implements a secure networking protocol. I would also like to do some minor changes in the OpenPlanet script and trackmania environments before I release it because sometimes I get a timeout from OpenPlanet, which crashes the script and silently messes up subsequent training samples. And finally I would like to "officially" launch our unofficial roborace competition :)

The version that is on the yb/dev branch implements secure networking already.

**Comment by NumseBacon** [comment]
so no new features for training/ai in 0.4.0? o.o

**Comment by yannbouteiller** [comment]
Nothing that does not exist yet, perhaps we can make support for training with full images instead of LIDARs in Trackmania more apparent, but I haven't managed to train something I am really satisfied with yet using this, the current AI goes faster than the LIDAR one but it hits a couple walls.

If you have suggestions in mind in terms of new training features they are welcome though !

**Comment by yannbouteiller** [comment]
Actually there is also a simple debugging tool that enables visualizing the gradient flow in each layer of the neural net on wandb. This part needs some love before it goes public though as it may otherwise flood the public wandb project with tons of PNGs.

**Comment by NumseBacon** [comment]
Sadly i dont know much about training AI in coding, so idk what to suggest

**Comment by yannbouteiller** [comment, marked as answer]
Another big feature of 0.4.0 is that we will move away from pytorch (in the sense that we will stop forcing people to use pytorch if they don't want to).

In particular, this will enable competitors to design all kinds of ActorModule which won't necessarily be subclasses of torch.nn.Module. This will also enable us to work with TensorFlow/Keras/Jax if we wish to.

**Comment by yannbouteiller** [comment]
`0.4.0` is up and running, I have uploaded it today on pypi :)

Amongst other things, it has a new default `config.json` that uses the `Full` environment instead of the `LIDAR` one, with hyperparameters that train a sensible CNN policy in a couple days.

It has also the competition tutorial that I recommend checking out for anyone who desire designing advanced things in TrackMania.

**Reply by NumseBacon** [reply]
Hey thanks for letting me know, sadly im having problems as the tmrl wont install the tmrldata folder :/

**Reply by yannbouteiller** [reply]
Really? That's weird, what happens? Is it just not there?

If in a python script you do:
```python
from pathlib import Path
print(Path.home())
```

It is supposed to print the folder where pip creates the `TmrlData` folder

**Reply by NumseBacon** [reply]
yeah theres just nothing

**Reply by yannbouteiller** [reply]
Heh... Strange. How are you installing the library, and on which OS?

**Reply by NumseBacon** [reply]
Im just doing what it says in [Installation](https://github.com/trackmania-rl/tmrl/blob/master/readme/Install.md) and
[Getting started](https://github.com/trackmania-rl/tmrl/blob/master/readme/get_started.md). Im on windows 10

**Reply by yannbouteiller** [reply]
What happens if you execute this script please?

```python
from pathlib import Path

HOME_FOLDER = Path.home()
TMRL_FOLDER = HOME_FOLDER / "TmrlData"

print(HOME_FOLDER.exists())
print(TMRL_FOLDER.exists())
```
